lib/puppet/network/http/connection.rb

```diff
@@ -189,6 +190,42 @@ def request_with_redirects(request, options)
             request.each do |header, value|
               current_request[header] = value
             end
+          when 429, 503
+            retry_after = current_response['Retry-After']
+
+            if retry_after.nil?
+              # Set return value and advance to next loop.
+              response = current_response
+              next
+            end
+
+            retry_after = begin
+                            Integer(retry_after)
+                          rescue TypeError, ArgumentError
+                            begin
+                              DateTime.rfc2822(retry_after)
+                            rescue ArgumentError
+                              Puppet.err(_('Received a %{status_code} response from the server, but the Retry-After header value of "%{retry_after}" could not be converted to an integer or RFC 2822 Date.') %
+                                         {status_code: current_response.code,
+                                          retry_after: retry_after.inspect})
+
+                              response = current_response
+                              next
+                            end
+                          end
+
+            retry_sleep = case retry_after
+                          when DateTime
+                            (retry_after.to_time - DateTime.now.to_time).to_i
+                          when Integer
+                            retry_after
+                          end
+
+            Puppet.warning(_('Received a %{status_code} response from the server. Sleeping for %{retry_after} seconds before retrying the request.') %
+                           {status_code: current_response.code,
+                            retry_after: retry_after})
+
+            ::Kernel.sleep(retry_sleep) if (retry_sleep > 0)
```

Comment:
Do you want to upper bounds the maximum sleep time so they agent doesn't hang if a server provides bad data / the headers are modified by an intermediary?

Reply:
I've waffled a bit about this and ended up leaving it out for the sake of simplicity. As it stands, there's a single upper limit that can be set on the Puppet Server side that caps the size of the Retry-After headers it sends and that's that.

But maybe using `Puppet[:runinterval]` as an upper bound would provide a measure of safety without being too confusing to users?

Reply:
Yeah, that might be a good compromise.  I'm just concerned that this sets a bit of a booby-trap that can't be escaped, whether that's due to:

* server-side misconfiguration where the retry sleep is set too long, agents grabs it as part of a failed request, then is stuck with the incorrect value until it expires
* the code malfunctions in some unexpected way
* something is done in a nefarious way to disable puppet from running / remediating for an extended period of time